File: src/main/java/wraith/fabricaeexnihilo/json/recipe/ToolRecipeJson.java
package wraith.fabricaeexnihilo.json.recipe;

import com.google.gson.*;
import wraith.fabricaeexnihilo.api.crafting.ItemIngredient;
import wraith.fabricaeexnihilo.api.crafting.Loot;
import wraith.fabricaeexnihilo.api.recipes.ToolRecipe;
import wraith.fabricaeexnihilo.json.BaseJson;

import java.lang.reflect.Type;
import java.util.stream.StreamSupport;

public final class ToolRecipeJson extends BaseJson<ToolRecipe> {

    private ToolRecipeJson() {}

    public static final ToolRecipeJson INSTANCE = new ToolRecipeJson();

    @Override
    public ToolRecipe deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        var obj = json.getAsJsonObject();
        return new ToolRecipe(
                ItemIngredient.fromJson(obj.get("ingredient")),
                StreamSupport.stream(obj.get("loot").getAsJsonArray().spliterator(), false).map(loot -> context.<Loot>deserialize(loot, LOOTABLE_TYPE_TOKEN)).toList()
        );
    }

    @Override
    public JsonElement serialize(ToolRecipe src, Type typeOfSrc, JsonSerializationContext context) {
        var obj = new JsonObject();
        obj.add("ingredient", src.ingredient().toJson());
        obj.add("loot", context.serialize(src.loots()));
        return obj;
    }

}

File: src/main/kotlin/exnihilofabrico/json/ToolRecipeJson.kt
package exnihilofabrico.json

import com.google.gson.*
import com.google.gson.reflect.TypeToken
import exnihilofabrico.api.crafting.ItemIngredient
import exnihilofabrico.api.crafting.Lootable
import exnihilofabrico.api.recipes.ToolRecipe
import java.lang.reflect.Type


object ToolRecipeJson: JsonDeserializer<ToolRecipe>, JsonSerializer<ToolRecipe> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): ToolRecipe {
        return ToolRecipe(
            ItemIngredient.fromJson(json.asJsonObject.get("ingredient"), context),
            json.asJsonObject.get("loot").asJsonArray.map { context.deserialize<Lootable>(it, object: TypeToken<Lootable>(){}.type) }.toMutableList()
        )
    }

    override fun serialize(src: ToolRecipe, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val obj = JsonObject()
        obj.add("ingredient", src.ingredient.toJson(context))
        obj.add("loot", context.serialize(src.lootables))
        return obj
    }

}
